recipes-extended/xen/files/xsm-for-vtd.patch

```diff
@@ -0,0 +1,259 @@
+SHORT DESCRIPTION:
+################################################################################
+Enable XSM control over device passthrough with separate policy controls for:
+* with and without an IOMMU available
+* an IOMMU with more capable isolation, in this case via interrupt remapping
+
+################################################################################
+LONG DESCRIPTION:
+################################################################################
+
+################################################################################
+CHANGELOG
+################################################################################
+
+################################################################################
+REMOVAL
+################################################################################
+Via upstreaming.
+
+################################################################################
+UPSTREAM PLAN
+################################################################################
+Yes.
+
+################################################################################
+INTERNAL DEPENDENCIES
+################################################################################
+OpenXT's XSM policy in xsm-policy.git
+
+################################################################################
+PATCHES
+################################################################################
+diff --git a/xen/xsm/flask/hooks.c b/xen/xsm/flask/hooks.c
+index 59905e2..bac29ad 100644
+--- a/xen/xsm/flask/hooks.c
++++ b/xen/xsm/flask/hooks.c
+@@ -20,6 +20,7 @@
+ #include <xen/errno.h>
+ #include <xen/guest_access.h>
+ #include <xen/xenoprof.h>
++#include <xen/iommu.h>
+ #ifdef HAS_PCI
+ #include <asm/msi.h>
+ #endif
+@@ -866,11 +867,41 @@ static int flask_map_domain_msi (struct domain *d, int irq, void *data,
+ #endif
+ }
+ 
++static u32 flask_iommu_resource_use_perm(void)
++{
++    /* This function defines the permission level required for allowing
++     * domains to use assigned devices.
++     *
++     * Interrupt remapping is an essential property for ensuring strict
++     * isolation of devices.
++     *
++     * If strict isolation is not available then check the platform permission
++     * to see whether the policy allows use of a less capable IOMMU.
++     */
++    struct av_decision avd;
++    u32 csid = domain_sid(current->domain);
++
++    /* Use the noaudit method for checking this permission because
++     * we're identifying which _other_ permission to check for.
++     * Failure here is sometimes expected and always acceptable so
++     * it is important not to log it to protect the logs.
++     */
++    if (iommu_enabled && (iommu_intremap ||
++                          !avc_has_perm_noaudit(csid, SECINITSID_XEN,
++                                                SECCLASS_IOMMU,
++                                                IOMMU__USE_INSECURE, &avd)))
```

Comment:
Assuming you want the permission per-domain, I think it would be clearer to just have three distinct use_with_* bits rather than using a "hidden" iommu__use_insecure bit:
- resource__use (possibly rename to resource__use_noiommu) - checked if (!iommu_enabled)
- resource__use_iommu_nointremap - checked if (iommu_enabled && !iommu_intremap)
- resource__use_iommu - checked if (iommu_enabled && iommu_intremap)

I also think that if resource__use is mapped to one of these, it should be mapped to resource__use_iommu (i.e. the only fully secure one), with the other two being "enable this to allow less-secure systems to still work" bits.

If this doesn't need to be a per-domain policy, the existing Xen "iommu=" command line argument can enforce the setting; systems that can't use intremap would already need to disable it in that parameter.